File: python/75. sort-colors.py
from collections import defaultdict
from typing import List


# lc sol start
class Solution:
    COLOR_TO_ID = {
        'RED': 0,
        'WHITE': 1,
        'BLUE': 2,
    }

    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        color_count = defaultdict(int)

        for n in nums:
            color_count[n] += 1

        acc = 0
        for idx in range(color_count[self.COLOR_TO_ID['RED']]):
            nums[acc + idx] = self.COLOR_TO_ID['RED']
        acc += color_count[self.COLOR_TO_ID['RED']]
        for idx in range(color_count[self.COLOR_TO_ID['WHITE']]):
            nums[acc + idx] = self.COLOR_TO_ID['WHITE']
        acc += color_count[self.COLOR_TO_ID['WHITE']]
        for idx in range(color_count[self.COLOR_TO_ID['BLUE']]):
            nums[acc + idx] = self.COLOR_TO_ID['BLUE']


# lc sol end


def main() -> None:
    test_cases = {
        'inputs': [([2, 0, 2, 1, 1, 0],), ([2, 0, 1],)],
        'outputs': [[0, 0, 1, 1, 2, 2], [0, 1, 2]],
    }

    solution = Solution()

    for idx, args in enumerate(test_cases['inputs']):
        solution.sortColors(*args)
        assert args[0] == test_cases['outputs'][idx], \
            f"expected output {test_cases['outputs'][idx]} for input {args}, got {args[0]}"

    print('all sample test cases passed successfully!')


if __name__ == '__main__':
    main()

File: cpp/75. sort-colors.cpp
//
// Created by Rutvik Patel on 2/8/24.
//

#include "include.h"
#include <unordered_map>

// solution
class Solution {
private:
    static std::unordered_map<std::string, int> COLOR_TO_ID;
public:
    void sortColors(std::vector<int>& nums) {
        std::unordered_map<int, int> color_count = {
            { COLOR_TO_ID["RED"], 0 },
            { COLOR_TO_ID["WHITE"], 0 },
            { COLOR_TO_ID["BLUE"], 0 },
        };

        for (auto n : nums)
            ++color_count[n];

        int idx = 0, acc = 0;
        while (idx < acc + color_count[COLOR_TO_ID["RED"]]) nums[idx++] = COLOR_TO_ID["RED"];
        acc = idx;
        while (idx < acc + color_count[COLOR_TO_ID["WHITE"]]) nums[idx++] = COLOR_TO_ID["WHITE"];
        acc = idx;
        while (idx < acc + color_count[COLOR_TO_ID["BLUE"]]) nums[idx++] = COLOR_TO_ID["BLUE"];
    }
};

std::unordered_map<std::string, int> Solution::COLOR_TO_ID = {
    { "RED", 0 },
    { "WHITE", 1 },
    { "BLUE", 2 },
};

// main
int main(__attribute__((unused)) int argc, __attribute__((unused)) char *argv[]) {
    std::vector<std::pair<std::vector<int>, std::vector<int>>> test_cases = {
        { { 2, 0, 2, 1, 1, 0 }, { 0, 0, 1, 1, 2, 2 } },
        { { 2, 0, 1 }, { 0, 1, 2 } },
    };

    Solution solution;

    for (auto& [args, ans] : test_cases) {
        solution.sortColors(args);
        assert_m(args == ans, "Failed!");
    }

    log("All testcases passed!");
}